skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldCommandLineStateTest.kt

```diff
@@ -0,0 +1,135 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold.run
+
+import com.google.common.truth.Truth.assertThat
+import com.google.container.tools.skaffold.SkaffoldExecutorService
+import com.google.container.tools.skaffold.SkaffoldExecutorSettings
+import com.google.container.tools.skaffold.SkaffoldProcess
+import com.google.container.tools.skaffold.message
+import com.google.container.tools.test.ContainerToolsRule
+import com.google.container.tools.test.TestService
+import com.google.container.tools.test.expectThrows
+import com.intellij.execution.ExecutionException
+import com.intellij.execution.RunnerAndConfigurationSettings
+import com.intellij.execution.configurations.RunConfigurationBase
+import com.intellij.execution.configurations.SearchScopeProvider
+import com.intellij.execution.runners.ExecutionEnvironment
+import com.intellij.util.ThrowableRunnable
+import io.mockk.CapturingSlot
+import io.mockk.every
+import io.mockk.impl.annotations.MockK
+import io.mockk.mockk
+import io.mockk.mockkStatic
+import io.mockk.slot
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.io.File
+
+/** Unit tests for [SkaffoldCommandLineState] */
+class SkaffoldCommandLineStateTest {
+    @get:Rule
+    val containerToolsRule = ContainerToolsRule(this)
+
+    private lateinit var skaffoldCommandLineState: SkaffoldCommandLineState
+
+    @MockK
+    private lateinit var mockExecutionEnvironment: ExecutionEnvironment
+    @MockK
+    private lateinit var mockRunnerSettings: RunnerAndConfigurationSettings
+    @MockK
+    @TestService
+    private lateinit var mockSkaffoldExecutorService: SkaffoldExecutorService
+
+    private val skaffoldSettingsCapturingSlot: CapturingSlot<SkaffoldExecutorSettings> = slot()
+
+    @Before
+    fun setUp() {
+        every {
+            mockExecutionEnvironment.runnerAndConfigurationSettings
+        } answers { mockRunnerSettings }
+
+        // CommandLineState calls this static method, needs to be mocked
+        mockkStatic(SearchScopeProvider::class)
+        every { SearchScopeProvider.createSearchScope(any(), any()) } answers { mockk() }
+
+        // Skaffold executor answer mocks
+        val mockSkaffoldProcess: SkaffoldProcess = mockk(relaxed = true)
+        every {
+            mockSkaffoldExecutorService.executeSkaffold(capture(skaffoldSettingsCapturingSlot))
+        } answers { mockSkaffoldProcess }
+    }
+
+    @Test
+    fun `unsupported run configuration type throws execution exception`() {
+        every { mockRunnerSettings.configuration } answers { mockk<RunConfigurationBase>() }
```

Comment:
> mockk<RunConfigurationBase>()

Is that return what is causing the exception to be thrown in `startProcess` (presumably because its not an `AbstractSkaffoldRunConfiguration`?)? I wonder if it would be cleaner to declare the mock at the class level with a nice name to help readability? Something like `nonSkaffoldRunConfiguration`

Reply:
You mean define it as a non-Skaffold configuration as a separate class, or inner class?

Reply:
Maybe I misunderstood what the code is doing:

It looks like you are returning a mock when `mockRunnerSettings.configuration` is called (the equivalent of mockito's `when(foo).thenReturn(mock(bar))`? I was suggesting that instead of inline mocking, do the mocking as an annotation at the top with a nicely named variable.

Reply:
ah, so you meant variable with a cleaner name to make sure it's obvious this is a source of the issues? Sounds good, let me do that.

Reply:
yep. also in there is the suggestion to stick to annotation based mocking exclusively. I'm not sure I have a good reasoning as to why (so I could be misguided here), but its just a carry over from a rule I followed using mockito.

Reply:
I can extract mock Skaffold dev configuration, it used in 2 tests, but this one is very local. I'd say they it would make it hard to read to put all possible one use mocks on top of the class, wdyt? It seems like a good name and a local instantiation can make it easier to follow the logic here.

Reply:
I'd say go with what you prefer since it ultimately isn't a huge deal.

Personally I prefer declaring all the mocks at the class level. It makes it easy to quickly see what is being mocked in the test class. I also find the mixing of two mocking styles a bit messy.  But I also understand your point about it being local..